# GLOBAL WEATHER DATA 
- This is an analysis project to understand the weather of countries in 2024

In [4]:
import pandas as pd # data handling
import matplotlib.pyplot as plt # visualization
import seaborn as sns # visualization

- Pandas for data manipulaton
- Matplotlib and seaborn for visualization

## Loading Dataset 

In [5]:
df=pd.read_csv("weather.csv")
df

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9


## Inspecting Data

In [6]:
df.head()

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.4,26.6,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.33,19.82,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.1,2.0,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.76,3.05,Africa/Algiers,1715849100,2024-05-16 09:45,23.0,73.4,Sunny,...,10.4,18.4,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.50,1.52,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.7,0.9,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.84,13.23,Africa/Luanda,1715849100,2024-05-16 09:45,26.0,78.8,Partly cloudy,...,183.4,262.3,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55


In [7]:
df.tail()

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9
136632,Zimbabwe,Harare,-17.8178,31.0447,Africa/Harare,1776666600,2026-04-20 08:30,20.0,68.1,Sunny,...,7.85,8.05,1,1,06:06,17:43,09:07,20:18,Waxing Crescent,9


In [8]:
df.sample(4)

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
19161,Tunisia,Tunis,36.8000,10.18,Africa/Tunis,1724415300,2024-08-23 13:15,31.1,88.0,Partly cloudy,...,5.100,9.900,1,1,05:44,19:00,21:12,09:53,Waning Gibbous,85
21144,Belize,Belmopan,17.2500,-88.77,America/Belize,1725365700,2024-09-03 06:15,25.2,77.4,Cloudy,...,0.900,1.100,1,1,05:42,18:06,06:06,18:39,Waxing Crescent,0
78870,San Marino,San Marino,43.9333,12.45,Europe/San_Marino,1750928400,2025-06-26 11:00,29.3,84.7,Sunny,...,17.529,36.702,2,2,05:29,20:57,06:00,22:25,Waxing Crescent,0
61530,Romania,Bucharest,44.4333,26.10,Europe/Bucharest,1743240600,2025-03-29 11:30,10.2,50.4,Overcast,...,24.420,26.640,2,3,06:02,18:39,05:55,18:55,New Moon,0


## Checking Data Types

In [9]:
df.dtypes

country                          object
location_name                    object
latitude                        float64
longitude                       float64
timezone                         object
last_updated_epoch                int64
last_updated                     object
temperature_celsius             float64
temperature_fahrenheit          float64
condition_text                   object
wind_mph                        float64
wind_kph                        float64
wind_degree                       int64
wind_direction                   object
pressure_mb                       int64
pressure_in                     float64
precip_mm                       float64
precip_in                       float64
humidity                          int64
cloud                             int64
feels_like_celsius              float64
feels_like_fahrenheit           float64
visibility_km                   float64
visibility_miles                  int64
uv_index                        float64


In [10]:
df.shape

(136633, 41)

In [11]:
df.columns

Index(['country', 'location_name', 'latitude', 'longitude', 'timezone',
       'last_updated_epoch', 'last_updated', 'temperature_celsius',
       'temperature_fahrenheit', 'condition_text', 'wind_mph', 'wind_kph',
       'wind_degree', 'wind_direction', 'pressure_mb', 'pressure_in',
       'precip_mm', 'precip_in', 'humidity', 'cloud', 'feels_like_celsius',
       'feels_like_fahrenheit', 'visibility_km', 'visibility_miles',
       'uv_index', 'gust_mph', 'gust_kph', 'air_quality_Carbon_Monoxide',
       'air_quality_Ozone', 'air_quality_Nitrogen_dioxide',
       'air_quality_Sulphur_dioxide', 'air_quality_PM2.5', 'air_quality_PM10',
       'air_quality_us-epa-index', 'air_quality_gb-defra-index', 'sunrise',
       'sunset', 'moonrise', 'moonset', 'moon_phase', 'moon_illumination'],
      dtype='object')

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136633 entries, 0 to 136632
Data columns (total 41 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   country                       136633 non-null  object 
 1   location_name                 136633 non-null  object 
 2   latitude                      136633 non-null  float64
 3   longitude                     136633 non-null  float64
 4   timezone                      136633 non-null  object 
 5   last_updated_epoch            136633 non-null  int64  
 6   last_updated                  136633 non-null  object 
 7   temperature_celsius           136633 non-null  float64
 8   temperature_fahrenheit        136633 non-null  float64
 9   condition_text                136633 non-null  object 
 10  wind_mph                      136633 non-null  float64
 11  wind_kph                      136633 non-null  float64
 12  wind_degree                   136633 non-nul

## INFORMATIONS

- Number of coloumns = 41
- Number of rows = 136633
- missing values = 0
- memory usage = 42.7+ MB
- sunrise,moonset and sunset times stored as object 
- last_update stored as object 

Data Types
-  float = 21
- integer = 9
- objects = 11

## Selecting Columns

- Selecting single columns

In [13]:
df['country']

0         Afghanistan
1             Albania
2             Algeria
3             Andorra
4              Angola
             ...     
136628      Venezuela
136629        Vietnam
136630          Yemen
136631         Zambia
136632       Zimbabwe
Name: country, Length: 136633, dtype: object

In [14]:
df[['sunrise','sunset']]

,sunrise,sunset
0,04:50,18:50
1,05:21,19:54
2,05:40,19:50
3,06:31,21:11
4,06:12,17:55
...,...,...
136628,06:14,18:39
136629,05:34,18:17
136630,05:46,18:18
136631,06:15,17:56


In [15]:
place=df[['country','location_name']]

In [16]:
place

,country,location_name
0,Afghanistan,Kabul
1,Albania,Tirana
2,Algeria,Algiers
3,Andorra,Andorra La Vella
4,Angola,Luanda
...,...,...
136628,Venezuela,Caracas
136629,Vietnam,Hanoi
136630,Yemen,Sanaa
136631,Zambia,Lusaka


## Filtering the data

In [17]:
df[df['temperature_celsius']==25]

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
31,Cameroon,Yaounde,3.8700,11.5200,Africa/Douala,1715849100,2024-05-16 09:45,25.0,76.9,Partly Cloudy,...,25.60,31.20,2,3,06:01,18:19,13:13,00:59,Waxing Gibbous,55
43,Cyprus,Nicosia,35.1700,33.3700,Asia/Famagusta,1715849100,2024-05-16 11:45,25.0,77.0,Partly cloudy,...,7.70,13.50,1,1,05:42,19:45,13:11,02:08,Waxing Gibbous,55
82,Israel,Jerusalem,31.7800,35.2300,Asia/Jerusalem,1715849100,2024-05-16 11:45,25.0,77.0,Partly cloudy,...,5.00,9.40,1,1,05:42,19:30,13:07,01:55,Waxing Gibbous,55
144,Rwanda,Kigali,-1.9500,30.0600,Africa/Kigali,1715849100,2024-05-16 10:45,25.0,77.0,Partly cloudy,...,4.20,6.30,1,1,05:55,17:57,13:02,00:36,Waxing Gibbous,55
172,Tanzania,Dodoma,-6.1800,35.7500,Africa/Dar_es_Salaam,1715849100,2024-05-16 11:45,25.0,77.0,Partly cloudy,...,1.90,3.40,1,1,06:38,18:28,13:43,01:08,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136101,Equatorial Guinea,Malabo,3.7500,8.7833,Africa/Malabo,1776493800,2026-04-18 07:30,25.0,77.0,Patchy light rain with thunder,...,6.65,9.75,1,1,06:18,18:30,06:54,19:33,Waxing Crescent,0
136389,Saint Lucia,Castries,14.0000,-61.0000,America/St_Lucia,1776579300,2026-04-19 02:15,25.0,77.0,Partly cloudy,...,8.55,13.35,1,1,05:48,18:18,07:22,20:48,Waxing Crescent,4
136390,Saint Vincent and the Grenadines,Kingstown,13.1333,-61.2167,America/St_Vincent,1776579300,2026-04-19 02:15,25.0,77.0,Partly cloudy,...,6.60,10.00,1,1,05:50,18:18,07:25,20:47,Waxing Crescent,4
136483,Democratic Republic of Congo,Kinshasa,-4.3000,15.3000,Africa/Kinshasa,1776666600,2026-04-20 07:30,25.0,77.1,Patchy rain nearby,...,14.75,14.95,1,2,05:58,17:58,08:42,20:55,Waxing Crescent,9


In [18]:
df[df['moon_phase']=='New Moon']

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
4285,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1717682400,2024-06-06 18:30,18.0,64.3,Patchy rain nearby,...,3.90,8.20,1,1,04:40,19:04,04:02,19:18,New Moon,0
4286,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1717682400,2024-06-06 16:00,32.0,89.6,Sunny,...,4.50,5.10,1,1,05:08,20:11,04:31,20:41,New Moon,0
4287,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1717682400,2024-06-06 15:00,32.0,89.6,Sunny,...,16.40,34.80,2,2,05:29,20:04,04:58,20:33,New Moon,0
4288,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1717682400,2024-06-06 16:00,22.5,72.6,Patchy rain nearby,...,2.20,3.40,1,1,06:17,21:29,05:43,22:03,New Moon,0
4289,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1717682400,2024-06-06 15:00,26.0,78.8,Partly cloudy,...,7.30,25.20,1,1,06:17,17:55,06:04,17:55,New Moon,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136196,Samoa,Apia,-13.8333,-171.7333,Pacific/Apia,1776493800,2026-04-18 19:30,27.4,81.3,Partly cloudy,...,4.65,6.75,1,1,06:33,18:19,06:50,18:40,New Moon,0
136207,Solomon Islands,Honiara,-9.4333,159.9500,Pacific/Guadalcanal,1776493800,2026-04-18 17:30,31.2,88.2,Patchy rain nearby,...,2.45,2.95,1,1,06:23,18:16,06:42,18:44,New Moon,0
136224,Tonga,Nuku`Aloia,-21.1333,-175.2000,Pacific/Tongatapu,1776493800,2026-04-18 19:30,24.1,75.4,Clear,...,6.15,9.45,1,1,06:53,18:27,07:14,18:43,New Moon,0
136229,Tuvalu,Funafuti,-8.5167,179.2167,Pacific/Funafuti,1776493800,2026-04-18 18:30,30.3,86.5,Partly cloudy,...,1.75,2.25,1,1,06:06,17:59,06:21,18:25,New Moon,0


In [19]:
df[df['moon_phase']=='Waxing Crescent']

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
4478,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1717768800,2024-06-07 18:30,17.4,63.3,Sunny,...,5.20,10.50,1,1,04:40,19:04,04:53,20:21,Waxing Crescent,0
4479,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1717768800,2024-06-07 16:00,31.3,88.3,Sunny,...,5.90,6.40,1,1,05:08,20:12,05:22,21:45,Waxing Crescent,0
4480,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1717768800,2024-06-07 15:00,35.0,95.0,Partly cloudy,...,20.00,47.40,2,2,05:29,20:05,05:50,21:36,Waxing Crescent,0
4481,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1717768800,2024-06-07 16:00,22.9,73.2,Sunny,...,3.90,13.60,1,1,06:17,21:29,06:32,23:06,Waxing Crescent,0
4482,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1717768800,2024-06-07 15:00,26.3,79.3,Partly cloudy,...,10.20,41.60,1,1,06:17,17:55,07:06,18:54,Waxing Crescent,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9


In [20]:
df[df['moon_phase']=='Waxing Gibbous']

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
132922,Vanuatu,Port Vila,-17.7333,168.3167,Pacific/Efate,1775025900,2026-04-01 17:45,28.1,82.6,Patchy rain nearby,...,8.55,13.65,1,1,05:53,17:49,17:09,04:49,Waxing Gibbous,97
132924,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1775025900,2026-04-01 13:45,26.0,78.8,Cloudy,...,27.85,29.35,2,3,05:50,18:11,17:35,05:06,Waxing Gibbous,98
132925,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1775025900,2026-04-01 09:45,20.0,67.9,Sunny,...,29.75,147.95,2,3,05:59,18:15,17:49,05:19,Waxing Gibbous,98
132926,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1775025900,2026-04-01 08:45,19.3,66.8,Cloudy,...,4.45,4.65,1,1,06:12,18:09,17:43,05:29,Waxing Gibbous,99


In [21]:
df[(df['condition_text']=='Sunny')&(df['temperature_celsius']<=25)]

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
15,Belarus,Minsk,53.9000,27.5700,Europe/Minsk,1715849100,2024-05-16 11:45,17.0,62.6,Sunny,...,1.60,1.70,1,1,05:05,21:08,12:59,03:13,Waxing Gibbous,55
22,Botswana,Gaborone,-24.6500,25.9100,Africa/Gaborone,1715849100,2024-05-16 10:45,22.0,71.6,Sunny,...,16.40,24.80,2,2,06:45,17:40,13:40,00:27,Waxing Gibbous,55
44,Czech Republic,Prague,50.0800,14.4700,Europe/Prague,1715849100,2024-05-16 10:45,18.0,64.4,Sunny,...,3.40,3.90,1,1,05:14,20:44,13:04,02:55,Waxing Gibbous,55
46,Denmark,Copenhagen,55.6700,12.5800,Europe/Copenhagen,1715849100,2024-05-16 10:45,16.0,60.8,Sunny,...,5.80,6.90,1,1,04:56,21:17,12:58,03:19,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136620,Uganda,Kampala,0.3156,32.5656,Africa/Kampala,1776666600,2026-04-20 09:30,23.1,73.6,Sunny,...,16.85,17.25,2,2,06:45,18:52,09:21,21:53,Waxing Crescent,8
136623,United Kingdom,London,51.5171,-0.1062,Europe/London,1776666600,2026-04-20 07:30,3.2,37.8,Sunny,...,14.35,18.85,1,2,05:54,20:05,06:54,No moonset,Waxing Crescent,9
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9


## Selecting rows - iloc method

In [22]:
df.iloc[5:12]

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
5,Antigua and Barbuda,Saint John's,17.12,-61.85,America/Antigua,1715849100,2024-05-16 04:45,26.0,78.8,Partly cloudy,...,1.2,4.5,1,1,05:36,18:32,13:05,01:14,Waxing Gibbous,55
6,Argentina,Buenos Aires,-34.59,-58.67,America/Argentina/Buenos_Aires,1715849100,2024-05-16 05:45,8.0,46.4,Clear,...,4.0,5.3,1,1,07:43,17:59,14:36,01:04,Waxing Gibbous,55
7,Armenia,Yerevan,40.18,44.51,Asia/Yerevan,1715849100,2024-05-16 12:45,19.0,66.2,Partly cloudy,...,0.8,0.9,1,1,05:45,20:12,13:17,02:31,Waxing Gibbous,55
8,Australia,Canberra,-35.28,149.22,Australia/Sydney,1715849100,2024-05-16 18:45,9.0,48.2,Clear,...,3.7,5.4,1,1,06:52,17:07,13:31,No moonset,Waxing Gibbous,55
9,Austria,Vienna,48.20,16.37,Europe/Vienna,1715849100,2024-05-16 10:45,16.0,60.8,Partly cloudy,...,3.7,4.4,1,1,05:14,20:29,13:00,02:42,Waxing Gibbous,55
10,Azerbaijan,Baku,40.40,49.88,Asia/Baku,1715849100,2024-05-16 12:45,17.0,62.6,Partly cloudy,...,1.9,2.2,1,1,05:23,19:51,12:54,02:10,Waxing Gibbous,55
11,Bahamas,Nassau,25.08,-77.35,America/Nassau,1715849100,2024-05-16 04:45,27.0,80.6,Partly cloudy,...,12.2,24.6,1,2,06:24,19:48,14:01,02:26,Waxing Gibbous,55


In [23]:
## Loc Method
df.loc[300,'sunset']

'18:50'

In [24]:
df

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9


## Rename Column

In [25]:
df.rename(columns={'location_name':'location','condition_text':'condition'},inplace=True)

In [26]:
df

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9


In [27]:
df['condition'].value_counts()

condition
Partly cloudy                                  39376
Sunny                                          39229
Patchy rain nearby                             11376
Partly Cloudy                                   8669
Overcast                                        7410
Clear                                           7200
Mist                                            5716
Light rain                                      4287
Light rain shower                               3133
Fog                                             1709
Cloudy                                          1633
Moderate or heavy rain with thunder             1090
Moderate rain                                   1042
Patchy light rain with thunder                   843
Light drizzle                                    678
Light snow                                       523
Patchy light drizzle                             455
Moderate or heavy rain shower                    287
Thundery outbreaks in nearby        

## Data Cleaning

### Clean Numerical columns with characters

In [28]:
df['last_updated'] = pd.to_datetime(df['last_updated'],
                                   errors='coerce')


df.info()
                  

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136633 entries, 0 to 136632
Data columns (total 41 columns):
 #   Column                        Non-Null Count   Dtype         
---  ------                        --------------   -----         
 0   country                       136633 non-null  object        
 1   location                      136633 non-null  object        
 2   latitude                      136633 non-null  float64       
 3   longitude                     136633 non-null  float64       
 4   timezone                      136633 non-null  object        
 5   last_updated_epoch            136633 non-null  int64         
 6   last_updated                  136633 non-null  datetime64[ns]
 7   temperature_celsius           136633 non-null  float64       
 8   temperature_fahrenheit        136633 non-null  float64       
 9   condition                     136633 non-null  object        
 10  wind_mph                      136633 non-null  float64       
 11  wind_kph     

In [29]:
df.drop(columns=['sunrise'],inplace=True)

In [30]:
df

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition,...,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15:00,26.6,79.8,Partly Cloudy,...,0.20,8.40,26.60,1,1,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45:00,19.0,66.2,Partly cloudy,...,0.10,1.10,2.00,1,1,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45:00,23.0,73.4,Sunny,...,13.40,10.40,18.40,1,1,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45:00,6.3,43.3,Light drizzle,...,0.20,0.70,0.90,1,1,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45:00,26.0,78.8,Partly cloudy,...,31.50,183.40,262.30,5,10,17:55,13:17,00:38,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30:00,17.5,63.6,Partly Cloudy,...,1.65,7.85,9.35,1,1,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30:00,28.3,82.9,Cloudy,...,21.05,36.25,39.45,2,4,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30:00,21.0,69.7,Sunny,...,2.05,12.85,51.65,1,2,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30:00,21.5,70.8,Sunny,...,2.05,5.55,5.85,1,1,17:56,09:13,20:35,Waxing Crescent,9


In [31]:
df.drop(columns=['sunset','moonrise','moonset','moon_phase','moon_illumination'],inplace=True)

In [32]:
df.drop(columns=['last_updated_epoch','temperature_fahrenheit','wind_mph','pressure_in','precip_in','gust_mph'],inplace=True)

In [33]:
df.drop(columns=['visibility_miles','feels_like_fahrenheit'],inplace=True)

In [34]:
df

,country,location,latitude,longitude,timezone,last_updated,temperature_celsius,condition,wind_kph,wind_degree,...,uv_index,gust_kph,air_quality_Carbon_Monoxide,air_quality_Ozone,air_quality_Nitrogen_dioxide,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,2024-05-16 13:15:00,26.6,Partly Cloudy,13.3,338,...,7.0,15.3,277.00,103.0,1.10,0.20,8.40,26.60,1,1
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,2024-05-16 10:45:00,19.0,Partly cloudy,11.2,320,...,5.0,18.4,193.60,97.3,0.90,0.10,1.10,2.00,1,1
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,2024-05-16 09:45:00,23.0,Sunny,15.1,280,...,5.0,22.3,540.70,12.2,65.10,13.40,10.40,18.40,1,1
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,2024-05-16 10:45:00,6.3,Light drizzle,11.9,215,...,2.0,13.7,170.20,64.4,1.60,0.20,0.70,0.90,1,1
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,2024-05-16 09:45:00,26.0,Partly cloudy,13.0,150,...,8.0,20.2,2964.00,19.0,72.70,31.50,183.40,262.30,5,10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,2026-04-20 02:30:00,17.5,Partly Cloudy,3.6,122,...,0.0,10.8,577.85,35.0,8.15,1.65,7.85,9.35,1,1
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,2026-04-20 13:30:00,28.3,Cloudy,17.6,151,...,5.5,20.5,351.85,35.0,34.45,21.05,36.25,39.45,2,4
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,2026-04-20 09:30:00,21.0,Sunny,7.9,135,...,6.5,9.1,149.85,58.0,5.85,2.05,12.85,51.65,1,2
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,2026-04-20 08:30:00,21.5,Sunny,13.0,58,...,1.6,17.8,160.85,40.0,2.85,2.05,5.55,5.85,1,1


In [35]:
df.shape

(136633, 27)

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136633 entries, 0 to 136632
Data columns (total 27 columns):
 #   Column                        Non-Null Count   Dtype         
---  ------                        --------------   -----         
 0   country                       136633 non-null  object        
 1   location                      136633 non-null  object        
 2   latitude                      136633 non-null  float64       
 3   longitude                     136633 non-null  float64       
 4   timezone                      136633 non-null  object        
 5   last_updated                  136633 non-null  datetime64[ns]
 6   temperature_celsius           136633 non-null  float64       
 7   condition                     136633 non-null  object        
 8   wind_kph                      136633 non-null  float64       
 9   wind_degree                   136633 non-null  int64         
 10  wind_direction                136633 non-null  object        
 11  pressure_mb  

### Handling Missing values

In [37]:
df.isnull().sum()

country                         0
location                        0
latitude                        0
longitude                       0
timezone                        0
last_updated                    0
temperature_celsius             0
condition                       0
wind_kph                        0
wind_degree                     0
wind_direction                  0
pressure_mb                     0
precip_mm                       0
humidity                        0
cloud                           0
feels_like_celsius              0
visibility_km                   0
uv_index                        0
gust_kph                        0
air_quality_Carbon_Monoxide     0
air_quality_Ozone               0
air_quality_Nitrogen_dioxide    0
air_quality_Sulphur_dioxide     0
air_quality_PM2.5               0
air_quality_PM10                0
air_quality_us-epa-index        0
air_quality_gb-defra-index      0
dtype: int64

In [38]:
df['temperature_celsius'].median()

23.9

In [39]:
df['temperature_celsius'].mode()

0    25.1
Name: temperature_celsius, dtype: float64

In [40]:
print(df.select_dtypes(include='number').mean())

latitude                          19.211173
longitude                         21.956700
temperature_celsius               21.268191
wind_kph                          12.897045
wind_degree                      168.797348
pressure_mb                     1014.064304
precip_mm                          0.134168
humidity                          66.661553
cloud                             39.976228
feels_like_celsius                22.120190
visibility_km                      9.507450
uv_index                           3.313816
gust_kph                          18.242712
air_quality_Carbon_Monoxide      459.169559
air_quality_Ozone                 57.997988
air_quality_Nitrogen_dioxide      14.974117
air_quality_Sulphur_dioxide       10.315061
air_quality_PM2.5                 24.149276
air_quality_PM10                  48.293201
air_quality_us-epa-index           1.697430
air_quality_gb-defra-index         2.601275
dtype: float64


In [41]:
print(df.select_dtypes(include='number').median())

latitude                          17.2500
longitude                         23.2361
temperature_celsius               23.9000
wind_kph                          10.8000
wind_degree                      161.0000
pressure_mb                     1014.0000
precip_mm                          0.0000
humidity                          72.0000
cloud                             30.0000
feels_like_celsius                25.1000
visibility_km                     10.0000
uv_index                           1.9000
gust_kph                          15.3000
air_quality_Carbon_Monoxide      293.7000
air_quality_Ozone                 55.0000
air_quality_Nitrogen_dioxide       5.6500
air_quality_Sulphur_dioxide        2.4050
air_quality_PM2.5                 14.1500
air_quality_PM10                  19.9800
air_quality_us-epa-index           1.0000
air_quality_gb-defra-index         2.0000
dtype: float64


In [42]:
c= df.select_dtypes(include='object').mode()

In [43]:
c

,country,location,timezone,condition,wind_direction
0,Bulgaria,Accra,Asia/Bangkok,Partly cloudy,E
1,NaN,Amman,NaN,NaN,NaN
2,NaN,Asmara,NaN,NaN,NaN
3,NaN,Baghdad,NaN,NaN,NaN
4,NaN,Bern,NaN,NaN,NaN
5,NaN,Bujumbura,NaN,NaN,NaN
6,NaN,Dakar,NaN,NaN,NaN
7,NaN,Kabul,NaN,NaN,NaN
8,NaN,Kyiv,NaN,NaN,NaN
9,NaN,Malabo,NaN,NaN,NaN


In [45]:
c=df.select_dtypes(include='object')
for i in c:
    print(i,df[i].mode())
    print()

country 0    Bulgaria
Name: country, dtype: object

location 0            Accra
1            Amman
2           Asmara
3          Baghdad
4             Bern
5        Bujumbura
6            Dakar
7            Kabul
8             Kyiv
9           Malabo
10       N'djamena
11             Nan
12           Sanaa
13            Suva
14        Tashkent
15           Tokyo
16        Valletta
17    Vatican City
18          Warsaw
Name: location, dtype: object

timezone 0    Asia/Bangkok
Name: timezone, dtype: object

condition 0    Partly cloudy
Name: condition, dtype: object

wind_direction 0    E
Name: wind_direction, dtype: object



### Handling duplicate

In [46]:
df.duplicated().sum()

np.int64(1)

In [47]:
df.dtypes

country                                 object
location                                object
latitude                               float64
longitude                              float64
timezone                                object
last_updated                    datetime64[ns]
temperature_celsius                    float64
condition                               object
wind_kph                               float64
wind_degree                              int64
wind_direction                          object
pressure_mb                              int64
precip_mm                              float64
humidity                                 int64
cloud                                    int64
feels_like_celsius                     float64
visibility_km                          float64
uv_index                               float64
gust_kph                               float64
air_quality_Carbon_Monoxide            float64
air_quality_Ozone                      float64
air_quality_N

In [48]:
print('Before:',df.shape)
df=df.drop_duplicates()
print('After:',df.shape)

Before: (136633, 27)
After: (136632, 27)


In [49]:
df.duplicated().sum()

np.int64(0)